### PR TITLE
Add demo for HKDF

### LINKED ISSUE #14120: Add an EVP demo for HKDF
It should be the current best practice (fetch), have proper error handling and resource allocation/deallocation and be in the style of a known answer test.

### PR DESCRIPTION
**Pull request by kleinmrk**
Fixes #14120

### CODE REVIEW

**Review by slontis — COMMENTED** [review]
Looks pretty good.
An entry should be added into the alphabetical list in demos/README.txt

kdf:

File: demos/kdf/hkdf.c [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+static unsigned char hkdf_okm[] = {
+    0x3c, 0xb2, 0x5f, 0x25, 0xfa, 0xac, 0xd5, 0x7a, 0x90, 0x43, 0x4f, 0x64,
+    0xd0, 0x36, 0x2f, 0x2a, 0x2d, 0x2d, 0x0a, 0x90, 0xcf, 0x1a, 0x5a, 0x4c,
+    0x5d, 0xb0, 0x2d, 0x56, 0xec, 0xc4, 0xc5, 0xbf, 0x34, 0x00, 0x72, 0x08,
+    0xd5, 0xb8, 0x87, 0x18, 0x58, 0x65
+};
+
+int main(int argc, char **argv)
```
Comment (slontis) [inline demos/kdf/hkdf.c:42]: A comment saying what HKDF is might be useful.

File: demos/kdf/hkdf.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+#include <stdio.h>
+#include <openssl/core_names.h>
+#include <openssl/crypto.h>
+#include <openssl/kdf.h>
+#include <openssl/obj_mac.h>
+#include <openssl/params.h>
+
+/* test vector from
```
Comment (slontis) [inline demos/kdf/hkdf.c:17]: Multi-line comments should look like this
````
/*
 * Test vector from
 * ...
 * /
 ````

File: demos/kdf/hkdf.c (hunk 1 above)
Comment (paulidale) [reply demos/kdf/hkdf.c:42]: Comments explaining each step would be better again.

File: demos/kdf/Makefile [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#
+# To run the demos when linked with a shared library (default):
+#
+#    LD_LIBRARY_PATH=../.. ./EVP_PKEY_EC_keygen
```
Comment (mattcaswell) [inline demos/kdf/Makefile:4]: EVP_PKEY_EC_keygen -> hkdf

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: demos/kdf/hkdf.c (hunk 1 above)
Comment (kleinmrk) [reply demos/kdf/hkdf.c:44]: I added the kdf entry in the [README.md](https://github.com/openssl/openssl/pull/15784/files#diff-c6ec795ee57468b7429abb6a588dd5d1e9169c080399b325751df19549c33fecR22) and also commented the important pieces in the code.

File: demos/kdf/hkdf.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    /* Create a context for the key derivation operation */
+    kctx = EVP_KDF_CTX_new(kdf);
+    if (kctx == NULL) {
+        fprintf(stderr, "EVP_KDF_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /* Set the underlying hash function used to derive the key */
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
```
Comment (paulidale) [inline demos/kdf/hkdf.c:74]: "SHA265" instead of SN_SHA256 for the demos.

**Review by paulidale — COMMENTED** [review]
Looks good, one small change.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for this contribution.

**Pull request closed without merging**